src/block/Hopper.php

```diff
@@ -92,7 +97,68 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$world = $this->position->getWorld();
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper) { $world->scheduleDelayedBlockUpdate($this->position, 1); return; }
+
+		$facingBlock = $this->getSide($this->facing);
+		$topBlock = $this->getSide(Facing::UP);
+		if(!$facingBlock instanceof HopperInteractable && !$topBlock instanceof Chest && !$topBlock instanceof Hopper) { $world->scheduleDelayedBlockUpdate($this->position, 1); return; }
+
+		$pullSuccess = $this->pull($tile);
+
+		if($pullSuccess && !$facingBlock instanceof HopperInteractable){
+			$world->scheduleDelayedBlockUpdate($this->position, 8);
+			return;
+		}
+
+		if(!$facingBlock instanceof HopperInteractable) { $world->scheduleDelayedBlockUpdate($this->position, 1); return; }
+		$updateTime = $facingBlock->push($tile) ? 8 : 1;
```

Comment:
Hoppers push before they pull as stated in the wiki: `Item pushes and pulls are processed in the same game tick, but pushes are processed before pulls.`